File: app/src/main/java/com/loksewa/aiapp/data/repository/AuthRepository.kt
package com.loksewa.aiapp.data.repository

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.loksewa.aiapp.data.local.UserDao
import com.loksewa.aiapp.data.local.UserEntity
import com.loksewa.aiapp.data.remote.LoksewaApiService
import com.loksewa.aiapp.data.remote.LoginRequest
import com.loksewa.aiapp.data.remote.RegisterRequest
import com.loksewa.aiapp.data.remote.AuthResponse
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.first
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TokenManager @Inject constructor(
    @ApplicationContext context: Context,
    private val dataStore: DataStore<Preferences>
) {
    private val legacyTokenKey = stringPreferencesKey("auth_token")
    private val securePreferences by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    suspend fun getToken(): String? {
        val encryptedToken = securePreferences.getString(SECURE_TOKEN_KEY, null)
        if (!encryptedToken.isNullOrBlank()) return encryptedToken

        val legacyToken = dataStore.data.first()[legacyTokenKey]
        if (!legacyToken.isNullOrBlank()) {
            saveToken(legacyToken)
            return legacyToken
        }
        return null
    }

    suspend fun saveToken(token: String) {
        securePreferences.edit().putString(SECURE_TOKEN_KEY, token).apply()
        dataStore.edit { it.remove(legacyTokenKey) }
    }

    suspend fun clearToken() {
        securePreferences.edit().remove(SECURE_TOKEN_KEY).apply()
        dataStore.edit { it.remove(legacyTokenKey) }
    }

    private companion object {
        const val SECURE_TOKEN_KEY = "auth_token"
    }
}

@Singleton
class AuthRepository @Inject constructor(
    private val apiService: LoksewaApiService,
    private val userDao: UserDao,
    private val tokenManager: TokenManager
) {
    suspend fun login(email: String, password: String): Result<AuthResponse> {
        return try {
            val response = apiService.login(LoginRequest(email, password))
            if (response.isSuccessful && response.body() != null) {
                val authResponse = response.body()!!
                tokenManager.saveToken(authResponse.token)
                
                // Cache user profile locally
                val user = authResponse.user
                userDao.insertUser(
                    UserEntity(
                        id = user.id,
                        email = user.email,
                        fullName = user.fullName,
                        role = user.role,
                        status = user.status,
                        createdAt = user.createdAt,
                        updatedAt = user.updatedAt,
                        lastLoginAt = user.lastLoginAt
                    )
                )
                Result.success(authResponse)
            } else {
                Result.failure(Exception(response.message().ifBlank { "Invalid email or password" }))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun register(email: String, password: String, fullName: String): Result<AuthResponse> {
        return try {
            val response = apiService.register(RegisterRequest(email, password, fullName))
            if (response.isSuccessful && response.body() != null) {
                val authResponse = response.body()!!
                tokenManager.saveToken(authResponse.token)
                
                // Cache user profile locally
                val user = authResponse.user
                userDao.insertUser(
                    UserEntity(
                        id = user.id,
                        email = user.email,
                        fullName = user.fullName,
                        role = user.role,
                        status = user.status,
                        createdAt = user.createdAt,
                        updatedAt = user.updatedAt,
                        lastLoginAt = user.lastLoginAt
                    )
                )
                Result.success(authResponse)
            } else {
                Result.failure(Exception(response.message().ifBlank { "Registration failed" }))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun logout(): Result<Unit> {
        return try {
            apiService.logout()
            tokenManager.clearToken()
            userDao.deleteAllUsers()
            Result.success(Unit)
        } catch (e: Exception) {
            // Even if network logout fails, clear local session
            tokenManager.clearToken()
            userDao.deleteAllUsers()
            Result.success(Unit)
        }
    }

    suspend fun getCachedUser(): UserEntity? {
        return userDao.getCurrentUser()
    }

    suspend fun isLoggedIn(): Boolean {
        return tokenManager.getToken() != null
    }
}

File: apps/mobile/lib/shared/repositories/auth_repository.dart
import '../models/user.dart';
import '../services/api_client.dart';

class AuthRepository {
  final ApiClient _api;

  const AuthRepository(this._api);

  Future<bool> get isSignedIn => _api.isSignedIn;

  Future<User> register({
    required String email,
    required String password,
    required String fullName,
  }) async {
    final response = await _api.post(
      '/api/auth/register',
      data: {
        'email': email.trim(),
        'password': password,
        'fullName': fullName.trim(),
      },
    );
    await _api.saveSession(response);
    final data = unwrap(response);
    return User.fromJson(data);
  }

  Future<User> login({
    required String email,
    required String password,
  }) async {
    final response = await _api.post(
      '/api/auth/login',
      data: {
        'email': email.trim(),
        'password': password,
      },
    );
    await _api.saveSession(response);
    final data = unwrap(response);
    return User.fromJson(data);
  }

  Future<bool> refresh(String refreshToken) async {
    try {
      final response = await _api.post(
        '/api/auth/refresh',
        data: {'refreshToken': refreshToken},
      );
      await _api.saveSession(response);
      return true;
    } catch (_) {
      await _api.clearSession();
      return false;
    }
  }

  Future<void> logout() async {
    try {
      await _api.post('/api/auth/logout');
    } catch (_) {
      // Clear session even if network call fails
    }
    await _api.clearSession();
  }

  Future<UserProfile> getMe() async {
    final response = await _api.get('/api/users/me');
    final data = unwrap(response);
    return UserProfile.fromJson(data);
  }
}
